custom_components/tahoma/climate_devices/atlantic_electrical_heater.py

```diff
@@ -1,43 +1,153 @@
+    HVAC_MODE_AUTO,
     HVAC_MODE_HEAT,
+    HVAC_MODE_OFF,
     PRESET_COMFORT,
     PRESET_ECO,
     PRESET_NONE,
 )
-from homeassistant.const import TEMP_CELSIUS
+from homeassistant.const import (
+    ATTR_TEMPERATURE,
+    EVENT_HOMEASSISTANT_START,
+    STATE_UNKNOWN,
+    TEMP_CELSIUS,
+)
+from homeassistant.core import callback
+from homeassistant.helpers.event import async_track_state_change
 
+from ..coordinator import TahomaDataUpdateCoordinator
 from ..tahoma_device import TahomaDevice
 
+_LOGGER = logging.getLogger(__name__)
+
 COMMAND_SET_HEATING_LEVEL = "setHeatingLevel"
+COMMAND_SET_TARGET_TEMPERATURE = "setTargetTemperature"
+COMMAND_SET_OPERATING_MODE = "setOperatingMode"
 
+CORE_OPERATING_MODE_STATE = "core:OperatingModeState"
+CORE_TARGET_TEMPERATURE_STATE = "core:TargetTemperatureState"
 IO_TARGET_HEATING_LEVEL_STATE = "io:TargetHeatingLevelState"
 
-PRESET_FREEZE = "Freeze"
-PRESET_FROST_PROTECTION = "frostprotection"
-PRESET_OFF = "off"
-
-MAP_PRESET_MODES = {
-    PRESET_OFF: PRESET_NONE,
-    PRESET_FROST_PROTECTION: PRESET_FREEZE,
-    PRESET_ECO: PRESET_ECO,
-    PRESET_COMFORT: PRESET_COMFORT,
+PRESET_BOOST = "boost"
+PRESET_COMFORT1 = "comfort 1"
+PRESET_COMFORT2 = "comfort 2"
+PRESET_FROST_PROTECTION = "Frost Protection"
+PRESET_SECURED = "Secured"
+
+PRESET_STATE_FROST_PROTECTION = "frostprotection"
+PRESET_STATE_OFF = "off"
+PRESET_STATE_ECO = "eco"
+PRESET_STATE_BOOST = "boost"
+PRESET_STATE_COMFORT = "comfort"
+PRESET_STATE_COMFORT1 = "comfort-1"
+PRESET_STATE_COMFORT2 = "comfort-2"
+
+# Map Home Assistant presets to TaHoma presets
+PRESET_MODE_TO_TAHOMA = {
+    PRESET_BOOST: PRESET_STATE_BOOST,
+    PRESET_COMFORT1: PRESET_STATE_COMFORT1,
+    PRESET_COMFORT2: PRESET_STATE_COMFORT2,
+    PRESET_COMFORT: PRESET_STATE_COMFORT,
+    PRESET_ECO: PRESET_STATE_ECO,
+    PRESET_FROST_PROTECTION: PRESET_STATE_FROST_PROTECTION,
+    PRESET_NONE: PRESET_STATE_OFF,
 }
 
-MAP_REVERSE_PRESET_MODES = {v: k for k, v in MAP_PRESET_MODES.items()}
+TAHOMA_TO_PRESET_MODE = {v: k for k, v in PRESET_MODE_TO_TAHOMA.items()}
 
-MAP_HVAC_MODES = {HVAC_MODE_HEAT: PRESET_COMFORT, HVAC_MODE_OFF: PRESET_OFF}
+# Map TaHoma HVAC modes to Home Assistant HVAC modes
+TAHOMA_TO_HVAC_MODE = {
+    "auto": HVAC_MODE_AUTO,
+    "basic": HVAC_MODE_HEAT,
+    "standby": HVAC_MODE_OFF,
+    "internal": HVAC_MODE_AUTO,
+    "off": HVAC_MODE_OFF,
+}
+
+HVAC_MODE_TO_TAHOMA = {v: k for k, v in TAHOMA_TO_HVAC_MODE.items()}
 
 
 class AtlanticElectricalHeater(TahomaDevice, ClimateEntity):
-    """Representation of TaHoma IO Atlantic Electrical Heater."""
+    """Representation of Atlantic Electrical Heater (With Adjustable Temperature Setpoint)."""
+
+    def __init__(self, device_url: str, coordinator: TahomaDataUpdateCoordinator):
+        """Init method."""
+        super().__init__(device_url, coordinator)
+
+        self._temp_sensor_entity_id = None
+        self._current_temperature = None
+
+    async def async_added_to_hass(self):
```

Comment:
This is exactly the same code as @tetienne uses in SomfyThermostat. Should we put this in a separate class and inherit from this code.

Reply:
I would say yes. It will avoid to duplicate code and maintenance.

Reply:
I am not 100% sure how to do this in the best way though.. It seems a bit complex with multiple inheritance and I am not able to test.. 
Let's do this in a follow up PR.

Reply:
Logged in https://github.com/iMicknl/ha-tahoma/issues/472.